### DISCUSSION TITLE
A fast way to pick all objects on screen?

### DISCUSSION THREAD

**Discussion by yanshjie** [visgl/deck.gl#7459] [Q&A]
we use pickobjects to do dynamic map labelling and we need to pick all objects on screen. Is there a faster way to do this than just use [pickobjects](https://deck.gl/docs/api-reference/core/deck#pickobjects) one time:
`deck.pickObjects({x: 0, y:0, width: viewport.width, height: viewport.height, layerIds: ...}`

![图片](https://user-images.githubusercontent.com/98957105/204691096-7ed77677-5c30-48db-b9d5-f4fc174158d0.png)

**Comment by ibgreen** [comment]
The answer is most likely no but here is some detail:

The thing that typically limits `pickObject(s)` performance is the readback of data from the GPU to the CPU. This synchronous readback stalls the GPU pipeline and introduces substantial overhead. Unfortunately, in WebGL this read back is done synchronously and there is no general way to fix this until we adopt WebGPU.

There is a WebGL extension that supports asynchronous readback in some cases but it doesn't work on Macs and all the deck.gl maintainers use MacBooks so there has been little interest to support it.

If you are doing multiple picks you might be able to gain speed by not doing a readback after every pick. But that would require you to dig deep into the picking code, this is not something the maintainers will do for you.

Alternatively, you could implement your own custom CPU picking algorithm, which depending on what type of data and how much data you have could be faster. 

As it stands, the main advantage pickObject(s) is really generality, not speed.

**Comment by Pessimistress** [comment]
> we use pickobjects to do dynamic map labelling and we need to pick all objects on screen

Without knowing more about your use case, I would like to share some generic observations:

- Processing geographic features by an arbitrary bounding box is going to be expensive, however you do it. You should avoid this expensive operation on every viewport change, otherwise it will kill your frame rate. You may want to consider the `TileLayer`, where you process the data for each tile only once.
- If you absolutely need to redo the query on the fly, consider an indexing library such as [rtree](https://github.com/leaflet-extras/RTree).
